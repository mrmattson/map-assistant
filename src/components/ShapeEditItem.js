import React from 'react'
import PropTypes from 'prop-types'
import R from 'ramda'
import inspector from 'schema-inspector'
import {
  FormGroup,
  FormControl,
  HelpBlock,
  ControlLabel,
  Button
} from 'react-bootstrap'

// import coordinateToString from '../helpers/coordinateToString'
// import { validCircleShape } from '../ShapeTools/circleShape'
import {
  NAUTICAL_MILES,
  METERS
} from '../helpers/constants'

const radiusSanitization = {
  type: 'number'
}

const radiusValidation = {
  type: 'number',
  gt: 0
}

function FieldGroup ({ id, label, help, validationState, ...props }) {
  return (
    <FormGroup controlId={id} validationState={validationState}>
      <ControlLabel>{label}</ControlLabel>
      <FormControl {...props} />
      {help && <HelpBlock>{help}</HelpBlock>}
    </FormGroup>
  )
}

function humanReadableUnits (units) {
  switch (units) {
    case NAUTICAL_MILES:
      return 'Nautical Miles'
    case METERS:
      return 'Meters'
    default:
      return 'Unknown'
  }
}

function ShapeEditItem ({ draftShape, selectedDraftShapes, index, onSelectionChange, onPropertyChange }) {
  const radiusValue = draftShape.properties.radius ? draftShape.properties.radius : ''
  const selected = R.contains(draftShape.id, selectedDraftShapes)

  return (
    <li className="list-group-item">
      <Button
        bsStyle={selected ? 'success' : 'default'}
        active={selected}
        onClick={() => onSelectionChange(draftShape.properties.id)}
      >{selected ? 'Selected' : 'Not Selected'}</Button>
      <FieldGroup
        id="formControlsText"
        type="text"
        label="Title"
        value={draftShape.properties.title}
        onChange={event => onPropertyChange({
          id: draftShape.properties.id,
          property: {
            title: event.target.value
          }
        })}
      />
      <FormGroup>
        <ControlLabel>Center</ControlLabel>
        <FormControl.Static>{draftShape.properties.unformattedCoordinates}</FormControl.Static>
      </FormGroup>
      <FieldGroup
        id="formControlsText"
        type="text"
        label="Radius"
        value={radiusValue}
        help={humanReadableUnits(draftShape.properties.radiusUnits)}
        onChange={event => onPropertyChange({
          id: draftShape.properties.id,
          property: {
            radius: inspector.sanitize(radiusSanitization, event.target.value).data
          }
        })}
        validationState={inspector.validate(radiusSanitization, radiusValue).valid
          ? null
          : 'error'
        }
      />
      <FormGroup>
        <ControlLabel>Original Text Input</ControlLabel>
        <FormControl
          componentClass="textarea"
          value={draftShape.properties.context}
          readOnly
        />
      </FormGroup>
    </li>
  )
}

// https://facebook.github.io/react/docs/typechecking-with-proptypes.html
ShapeEditItem.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  draftShape: PropTypes.object.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  selectedDraftShapes: PropTypes.array.isRequired,
  index: PropTypes.number.isRequired,
  onSelectionChange: PropTypes.func.isRequired,
  onPropertyChange: PropTypes.func.isRequired
}

export default ShapeEditItem
