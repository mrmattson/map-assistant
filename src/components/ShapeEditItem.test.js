/* eslint-env jest */

import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

import ShapeEditItem from './ShapeEditItem'

const draftShapeExample = {
  type: 'Feature',
  geometry: {
    type: 'Polygon',
    coordinates: [[[0, 1], [1, 0], [1, 1], [0, 1]]]
  },
  id: 'abcdefg',
  properties: {
    id: 'abcdefg',
    shape: 'Circle',
    title: '',
    description: '',
    stroke: '#000000',
    'stroke-width': 1,
    fill: '#000000',
    'fill-opacity': 0.5,
    radius: 1,
    radiusUnits: 'nauticalMiles',
    numSides: 12,
    centerCoordinates: [0, 1]
  }
}

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(
    <ShapeEditItem
      index={2}
      draftShape={draftShapeExample}
      selectedDraftShapes={[]}
      onSelectionChange={function onSelectionChange () { return 0 }}
      onPropertyChange={function onPropertyChange () { return 0 }}
    />,
    div
  )
})

it('clean snapshot', () => {
  const component = renderer.create(
    <ShapeEditItem
      index={2}
      draftShape={draftShapeExample}
      selectedDraftShapes={[]}
      onSelectionChange={function onSelectionChange () { return 0 }}
      onPropertyChange={function onPropertyChange () { return 0 }}
    />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
