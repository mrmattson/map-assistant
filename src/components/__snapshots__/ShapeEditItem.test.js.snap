exports[`test clean snapshot 1`] = `
<li
  className="list-group-item">
  <button
    className="btn btn-default"
    disabled={false}
    onClick={[Function]}
    type="button">
    Not Selected
  </button>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor="formControlsText">
      Title
    </label>
    <input
      className="form-control"
      id="formControlsText"
      onChange={[Function]}
      type="text"
      value="" />
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Center
    </label>
    <p
      className="form-control-static" />
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor="formControlsText">
      Radius
    </label>
    <input
      className="form-control"
      id="formControlsText"
      onChange={[Function]}
      type="text"
      value={1} />
    <span
      className="help-block">
      Nautical Miles
    </span>
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      Original Text Input
    </label>
    <textarea
      className="form-control"
      id={undefined}
      readOnly={true}
      type={undefined}
      value={undefined} />
  </div>
</li>
`;
